File: android/app/src/main/java/com/notesapp/data/model/Tag.kt
package com.notesapp.data.model

import com.google.gson.annotations.SerializedName

/** Represents a tag that can be attached to notes. */
data class Tag(
    val id: String,
    val name: String,
    val color: String,
    val createdAt: String,
    val updatedAt: String,
)

/** Tag with associated note count from the API. */
data class TagWithNoteCount(
    val id: String,
    val name: String,
    val color: String,
    val createdAt: String,
    val updatedAt: String,
    @SerializedName("_count") val count: NoteCount,
) {
    data class NoteCount(val notes: Int)
}

data class CreateTagInput(
    val name: String,
    val color: String,
)

data class UpdateTagInput(
    val name: String? = null,
    val color: String? = null,
)

data class AttachTagsInput(
    val tagIds: List<String>,
)

/** Wrapper for tags list API response. */
data class TagsApiResponse(
    val data: TagsData,
) {
    data class TagsData(val tags: List<TagWithNoteCount>)
}

/** Wrapper for single tag API response. */
data class TagApiResponse(
    val data: Tag,
)

/** Wrapper for generic success API response. */
data class MessageResponse(
    val data: MessageData,
) {
    data class MessageData(val message: String)
}

File: ios/NotesApp/Models/Tag.swift
import Foundation

/// Represents a tag that can be attached to notes.
struct Tag: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let name: String
    let color: String
    let createdAt: String
    let updatedAt: String
}

/// Tag with associated note count from the API.
struct TagWithNoteCount: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let name: String
    let color: String
    let createdAt: String
    let updatedAt: String
    let count: NoteCount

    enum CodingKeys: String, CodingKey {
        case id, name, color, createdAt, updatedAt
        case count = "_count"
    }

    struct NoteCount: Codable, Hashable, Sendable {
        let notes: Int
    }
}

struct CreateTagInput: Codable, Sendable {
    let name: String
    let color: String
}

struct UpdateTagInput: Codable, Sendable {
    let name: String?
    let color: String?
}

struct AttachTagsInput: Codable, Sendable {
    let tagIds: [String]
}

struct TagsResponse: Codable, Sendable {
    let data: TagsData

    struct TagsData: Codable, Sendable {
        let tags: [TagWithNoteCount]
    }
}

struct TagResponse: Codable, Sendable {
    let data: Tag
}
